File: ex/uri1610.bkp.java
import java.io.InputStreamReader;
import java.io.BufferedReader;
import java.io.IOException;
import java.util.List;
import java.util.ArrayList;

public class uri1610 {
  static InputStreamReader ir = new InputStreamReader(System.in);
  static BufferedReader in = new BufferedReader(ir);

  public static final void main(String[] args) throws IOException {
    int numberCases = Integer.parseInt(in.readLine());
    List path;
    String[] line;
    long past, future;
    int dom, sub;
    int doc, fil;
    int docs, rels;
    boolean[][] ships;
    boolean outlet;

    for (int c = 0; c < numberCases; c++) {
      past = System.nanoTime();

      /* read data */
      line = in.readLine().split(" ");
      docs = Integer.parseInt(line[0]);
      rels = Integer.parseInt(line[1]);
      ships = new boolean[docs][docs];

      for (int r = 0; r < rels; ++r) {
        line = in.readLine().split(" ");
        dom = Integer.parseInt(line[0]) - 1;
        sub = Integer.parseInt(line[1]) - 1;
        ships[dom][sub] = true;
      }

      /* process data */
      outlet = false;
      for (doc = 0; doc < docs && !outlet; ++doc) {
        path = new ArrayList<Integer>();
        for (fil = 0; fil < docs; ++fil)
          if (ships[doc][fil] == true) path.add(fil);
        for (int i = 0; i < path.size(); ++i) {
          dom = (int) path.get(i);
          for (sub = 0; sub < docs; ++sub)
            if (ships[dom][sub] == true && path.indexOf(sub) < 0)
              path.add(sub);
        }
        if (path.indexOf(doc) >= 0) outlet = true;
      }

      /* write data */
      if (outlet)
        System.out.println("SIM");
      else
        System.out.println("NAO");

      future = System.nanoTime();
      System.out.printf("  %.2fms\n", (future-past)/1000000.0);
    }
  }
}

File: ex/uri1610.bkp.cpp
#include <iostream>
#include <stdlib.h>
using namespace std;

bool result;
int docs;
int rels;
int past;
int future;
bool** ships;

void alert()
{
  cout << "!!!" << endl;
}

void inlet()
{
  int dom;
  int sub;

  cin >> docs >> rels;
  free(ships);
  ships = (bool**) malloc(docs*sizeof(bool*));
  for (int s = 0; s < docs; ++s)
  {
    ships[s] = (bool*) malloc(sizeof(bool)*docs);
    for (int i = 0; i < docs; ++i)
      ships[s][i] = false;
  }

  for (int r = 0; r < rels; ++r)
  {
    cin >> dom >> sub;
    ships[dom][sub] = true;
  }
}

void process()
{
  bool* path;
  int dom;
  int sub;
  int doc;
  int fil;
  int id;
  result = false;

  for (doc = 0; doc < docs && !result; ++doc)
  {
    path = (bool*) malloc(sizeof(bool)*docs);
    for (int i = 0; i < docs; ++i)
      path[i] = false;

    for (fil = 0; fil < docs; ++fil)
      if (ships[doc][fil]) path[fil] = true;

    for (id = 0; id < docs; ++id)
    {
      if (path[id] && id != doc) {
        dom = id;
        for (sub = 0; sub < docs; ++sub)
          if (ships[dom][sub] && !path[sub])
            path[sub] = true;
      }
    }

    if (path[doc])
      result = true;
    free(path);
  }
}

void outlet()
{
  if (result) cout << "SIM";
  else cout << "NAO";
  cout << endl;
}

main()
{
  int nc;

  cin >> nc;
  for (int c = 0; c < nc; c++)
  {
    inlet();
    process();
    outlet();
  }
}
